Native target

Him188 opened the pull request:
mirai-core native target
使用纯 Kotlin 实现
成品可以由 Kotlin/Native 依赖, 也可以生成动态链接库

部分现有 common API 在 native 平台不可用, 如 File, 将会直接删除. 部分已弃用很久的 API 也不会在 native 平台支持, 如 RemoteFile, Voice.

由于 Kotlin 协程不支持部分平台 x86 目标, 如 mingwx86, 故只支持这些平台的 x64 目标.

项目多平台架构现在是:
- common — 全平台通用, 绝大部分代码都在 common
   - jvmBase — JVM 通用
     - jvm — 桌面 JVM, 原 API 与 ABI 保留不变
     - android — Android VM, 原 API 与 ABI 保留不变
  - native — native 平台通用
     - mingwx64
     - unix
       - macosarm64
       - macosx64
       - linuxx64

在这之后也能较容易地支持 JS 目标, 不过 JS 目标应该不太有用途.

- [x] 迁移到多平台架构
- [x] 修复 JVM 目标测试
- [x] 实现 native 目标
   - [x] ~添加 Rust 编译和链接~
   - [x] 测试 MiraiFile on mingwx64
   - [x] 测试 MiraiFile on linuxx64
   - [x] ~MessageSerializers~ *拆分为 #2159*
   - [x] shadow ktor on JVM
- [x] actions
- [x] 检查 publishing

Him188 committed:
Configure HMPP project structure

Him188 committed:
Commonize projects: mirai-core series, and mirai-console-compiler-annotations

Him188 committed:
Use ktor-io to replace kotlinx-io

Him188 committed:
Implement multiplatform modules

AdoptOSS commented:
可能有用的信息：
考虑到Mirai项目庞大而复杂的多线程架构，建议基于新的 Native GC 构建

> The new Kotlin/Native memory manager lifts restrictions on object-sharing between threads. It also provides leak-free concurrent programming primitives that are safe and don’t require any special management or annotations.
> 
> The new memory manager will become the default one in future versions, so we encourage you to try it now. [Learn more](https://blog.jetbrains.com/kotlin/2021/08/try-the-new-kotlin-native-memory-manager-development-preview/) about the new memory manager and explore demo projects, or jump right to the [migration instructions](https://github.com/JetBrains/kotlin/blob/master/kotlin-native/NEW_MM.md) to try it yourself.

LaoLittle commented:
> 考虑到Mirai项目庞大而复杂的多线程架构，建议基于新的 Native GC 构建

以及coroutines-native-mt （）

AdoptOSS commented:
> 以及coroutines-native-mt （）

不，`native-mt`是在老内存模型下的残缺的多线程支持，new mm下直接用新版的普通版本就行

LaoLittle commented:
> > 以及coroutines-native-mt （）
> 
> 不，`native-mt`是在老内存模型下的残缺的多线程支持，new mm下直接用新版的普通版本就行

这样的么，我还以为mt就是多线程支持

Him188 commented:
现在全项目已经可以编译, 正在修复因迁移多平台导致的 jvm 目标测试失败. ~注意, commits 记录前大半部分是 rebase 自 #1997, 1997 合并会在近期并入 dev.~

----
启用了新 experimental 内存管理

Him188 committed:
Improve setState stability

LaoLittle commented:
‘currentTimeMillis’的话kotlin有个内置函数’getTimeMillis’，无需手动调用c库

Him188 commented:
> ‘currentTimeMillis’的话kotlin有个内置函数’getTimeMillis’，无需手动调用c库

那个是 ktor-utils 的东西，不是标准库，指不定哪天它就没了，最好不能依赖它

LaoLittle commented:
> 那个是 ktor-utils 的东西，不是标准库，指不定哪天它就没了，最好不能依赖它

在jvm里确实，不过在native上是`kotlin.system`

还有时间类工具其实有一套跨平台kotlin库

详见 https://github.com/Kotlin/kotlinx-datetime

LaoLittle commented:
还有就是native的动态库链接可能会有问题（在linux主要是konan所使用的工具链glibc版本过低导致的，windows环境下未知）

[KT-52510](https://youtrack.jetbrains.com/issue/KT-52510)
[KT-43501](https://youtrack.jetbrains.com/issue/KT-43501)
[KT-47061](https://youtrack.jetbrains.com/issue/KT-47061)

Him188 commented:
我知道有 kotlinx-datetime，但是增加一个库就增加了体积，对 mirai 只需要 current time 来说是没必要的。linker 目前没有遇到问题。

LaoLittle commented:
> 我知道有 kotlinx-datetime，但是增加一个库就

好的

WetABQ commented:
> 我知道有 kotlinx-datetime，但是增加一个库就增加了体积，对 mirai 只需要 current time 来说是没必要的。linker 目前没有遇到问题。

`kotlinx-datetime` 目前还没有正式发布，并且它将使用的 Kotlin 标准库升级到了 `1.7.0-Beta`，会导致项目引入多个不同版本状态下的 Kotlin 标准库

Him188 committed:
HMPP

Him188 commented:
由于要支持数个目标平台，和一些 Kotlin 现阶段交互限制，我打算改为纯 Kotlin 实现。将会提供 C 交互 API，有可能未来也会提供 Rust API。

Him188 committed:
Ensure for all MessageChain subclasses, `equals`, `hashCode` give consistent results.

Him188 committed:
Migrate `-Xopt-in` to `-opt-in`

Him188 committed:
Implement mirai-core for native

Him188 committed:
Rewrite MessageSerializers, register serializers in each MessageProtocol, implement for native targets

Him188 committed:
Add native test entrance

Him188 commented:
现在在 macos 可以登录
<img width="1380" alt="image" src="https://user-images.githubusercontent.com/12100985/171513154-2b8e97ba-731c-47f4-b0bc-279375572c10.png">

Him188 committed:
Use Zlib for compression/decompression on native

AdoptOSS commented on mirai-core-utils/src/mingwX64Main/kotlin/MiraiFileImpl.kt:
```diff
@@ -0,0 +1,174 @@
+
+        memScoped {
+            val b = alloc<WINBOOLVar>()
+            CheckNameLegalDOS8Dot3A(absolutePath, nullPtr(), 0, nullPtr(), b.ptr)
```
Use unicode version api `CheckNameLegalDOS8Dot3W`

AdoptOSS replied:
> This function can be used to determine whether or not a file name can be passed to a 16-bit Windows-based application or an MS-DOS-based application.

这里似乎不应该使用它，Mirai创建的文件并不必须要可以在 16-bit 环境下可用

AdoptOSS commented on mirai-core-utils/src/mingwX64Main/kotlin/MiraiFileImpl.kt:
```diff
@@ -0,0 +1,174 @@
+
+    override fun exists(): Boolean = getFileAttributes() != INVALID_FILE_ATTRIBUTES
+
+    private fun getFileAttributes(): DWORD = memScoped { GetFileAttributesA(absolutePath) }
```
Use unicode version api `GetFileAttributesW`

AdoptOSS commented on mirai-core-utils/src/mingwX64Main/kotlin/MiraiFileImpl.kt:
```diff
@@ -0,0 +1,174 @@
+    override fun createNewFile(): Boolean {
+        memScoped {
+            // https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea
+            val handle = CreateFileA(
```
Use unicode version api `CreateFileW`

AdoptOSS commented on mirai-core-utils/src/mingwX64Main/kotlin/MiraiFileImpl.kt:
```diff
@@ -0,0 +1,174 @@
+
+    override fun delete(): Boolean {
+        return if (isFile) {
+            DeleteFileA(absolutePath) == 0
```
Use unicode version api `DeleteFileW`

AdoptOSS commented on mirai-core-utils/src/mingwX64Main/kotlin/MiraiFileImpl.kt:
```diff
@@ -0,0 +1,174 @@
+        return if (isFile) {
+            DeleteFileA(absolutePath) == 0
+        } else {
+            RemoveDirectoryA(absolutePath) == 0
```
Use unicode version api `RemoveDirectoryW`

AdoptOSS commented on mirai-core-utils/src/mingwX64Main/kotlin/MiraiFileImpl.kt:
```diff
@@ -0,0 +1,174 @@
+    override fun mkdir(): Boolean {
+        memScoped {
+            val v = alloc<_SECURITY_ATTRIBUTES>()
+            return CreateDirectoryA(absolutePath, v.ptr) == 0
```
Use unicode version api `CreateDirectoryW`

AdoptOSS commented on mirai-core-utils/src/mingwX64Main/kotlin/MiraiFileImpl.kt:
```diff
@@ -0,0 +1,174 @@
+        actual fun getWorkingDir(): MiraiFile {
+            val path = memScoped {
+                ByteArray(PATH_MAX).usePinned {
+                    getcwd(it.addressOf(0), it.get().size.convert())
```
Use unicode version api `GetCurrentDirectoryW`

Him188 committed:
Introduce higher performance compression/decompression methods for different purposes for common

Him188 commented:
![image](https://user-images.githubusercontent.com/12100985/172016775-73dd83d0-7f33-4355-afe4-4104619e9d10.png)
现在 macOS 下全部测试通过

Him188 commented:
@AdoptOSS 谢谢, 之后测试 Windows 平台的时候会去修复 , 如果你想 review 那我之后就不 force push 了

Him188 committed:
Configure GitHub Actions for native

Him188 committed:
Use different types for TestState

Him188 committed:
Improve core network stability:

Close network and do logout only if networkInitialized

core, network: Guard changing state

core, network: Protect state overriding when setting a state

Always close NetworkHandlerSupport even if CommonNetworkHandler is in CLOSED state

Him188 committed:
Cancel TestEventDispatcher jobs after test

Him188 committed:
Broadcast BotOfflineEvent.Active after state OK -> CLOSED

AdoptOSS commented on buildSrc/src/main/kotlin/HmppConfigure.kt:
```diff
@@ -0,0 +1,507 @@
+    }
+}
+
+private fun Project.configureNativeInterop(
```
Clean up

Him188 replied:
不确定以后是否还会用它, 就留着了, 已经在我本地注释掉了

AdoptOSS commented on mirai-console/backend/mirai-console/test/command/LoginCommandTest.kt:
```diff
@@ -32,7 +33,7 @@ import kotlin.test.assertEquals
 
     @Test
-    suspend fun `login with provided password`() {
+    fun `login with provided password`() = runBlocking {
```
Use `kotlinx-coroutines-test`

AdoptOSS commented on mirai-console/backend/mirai-console/test/command/LoginCommandTest.kt:
```diff
@@ -52,7 +53,7 @@ internal class LoginCommandTest : AbstractCommandTest() {
 
     @Test
-    suspend fun `login with saved plain password`() {
+    fun `login with saved plain password`() = runBlocking {
```
Use `kotlinx-coroutines-test`

AdoptOSS commented on mirai-console/backend/mirai-console/test/command/LoginCommandTest.kt:
```diff
@@ -81,7 +82,7 @@ internal class LoginCommandTest : AbstractCommandTest() {
 
     @Test
-    suspend fun `login with saved md5 password`() {
+    fun `login with saved md5 password`() = runBlocking {
```
Use `kotlinx-coroutines-test`

AdoptOSS commented on mirai-core-api/src/androidMain/kotlin/package.kt:
```diff
@@ -0,0 +1,10 @@
+ * https://github.com/mamoe/mirai/blob/dev/LICENSE
+ */
+
+package net.mamoe.mirai
```
好奇这个文件的用途

Him188 replied:
解决 IDE 在复制粘贴/移动文件时错误地删除包名的问题

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/Event.kt:
```diff
@@ -152,6 +154,7 @@ public interface CancellableEvent : Event {
  * [EventChannel.filter] 和 [Listener.onEvent] 时产生的异常只会由监听方处理.
  */
 @JvmBlockingBridge
+@Suppress("TOP_LEVEL_FUNCTIONS_NOT_SUPPORTED") // compiler bug
```
Details?
特殊注释
```suggestion
@Suppress("TOP_LEVEL_FUNCTIONS_NOT_SUPPORTED") // FIXME compiler bug
```

Him188 replied:
是 KJBB 的 IDE 问题, 不用特地让人注意

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/events/friend.kt:
```diff
@@ -80,7 +84,7 @@ public data class NewFriendRequestEvent @MiraiInternalApi public constructor(
     public val fromNick: String,
 ) : BotEvent, Packet, AbstractEvent(), FriendInfoChangeEvent {
     @JvmField
```
对于atomicfu而言，在JVM下默认就会转换了field并调用xxxUpdater
```suggestion
```

Him188 replied:
它确实会 transform, 改的时候没注意

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/events/group.kt:
```diff
@@ -350,7 +352,7 @@ public data class BotInvitedJoinGroupRequestEvent @MiraiInternalApi constructor(
     public val invitor: Friend? get() = this.bot.getFriend(invitorId)
 
     @JvmField
```
```suggestion
```

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/events/group.kt:
```diff
@@ -400,7 +402,7 @@ public data class MemberJoinRequestEvent @MiraiInternalApi constructor(
 
     @JvmField
```
```suggestion
```

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/events/group.kt:
```diff
@@ -400,7 +402,7 @@ public data class MemberJoinRequestEvent @MiraiInternalApi constructor(
 
     @JvmField
     @PublishedApi
-    internal val responded: AtomicBoolean = AtomicBoolean(false)
+    internal val responded: AtomicBoolean = atomic(false)
```
atomicfu通过自动修改内部代码实现，可能无法被安全的`published`
考虑使用委托 (due to Kotlin/kotlinx.atomicfu#186, 可能需要手动委托)

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/events/group.kt:
```diff
@@ -442,7 +444,7 @@ public data class MemberJoinRequestEvent @MiraiInternalApi constructor(
         @JvmStatic
         @JvmName("copy\$default") // avoid being mangled
-        fun `copy$default`(
+        fun copy_default(
```
既然是为了jvm兼容性存在的，just move it into jvmMain

Him188 replied:
那会需要把这整个 class 变成 expect/actual, 会增加复杂度, 在考虑之后还是留 common 好

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/select.kt:
```diff
@@ -404,7 +407,7 @@ public abstract class MessageSelectBuilderUnit<M : MessageEvent, R> @PublishedAp
     @JvmName("invoke-RNyhSv4")
     @Deprecated("For binary compatibility", level = DeprecationLevel.HIDDEN)
-    public fun MessageSelectionTimeoutChecker.invoke000(block: suspend () -> R): Void? {
+    public fun MessageSelectionTimeoutChecker.invoke000(block: suspend () -> R): Nothing? {
```
move into jvmMain

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/select.kt:
```diff
@@ -417,7 +420,7 @@ public abstract class MessageSelectBuilderUnit<M : MessageEvent, R> @PublishedAp
     @JvmName("reply-RNyhSv4")
     @Deprecated("For binary compatibility", level = DeprecationLevel.HIDDEN)
-    public infix fun MessageSelectionTimeoutChecker.reply000(block: suspend () -> Any?): Void? {
+    public infix fun MessageSelectionTimeoutChecker.reply000(block: suspend () -> Any?): Nothing? {
```
move into jvmMain

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/select.kt:
```diff
@@ -430,7 +433,7 @@ public abstract class MessageSelectBuilderUnit<M : MessageEvent, R> @PublishedAp
     @JvmName("reply-sCZ5gAI")
     @Deprecated("For binary compatibility", level = DeprecationLevel.HIDDEN)
-    public infix fun MessageSelectionTimeoutChecker.reply000(message: String): Void? {
+    public infix fun MessageSelectionTimeoutChecker.reply000(message: String): Nothing? {
```
move into jvmMain

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/select.kt:
```diff
@@ -443,7 +446,7 @@ public abstract class MessageSelectBuilderUnit<M : MessageEvent, R> @PublishedAp
     @JvmName("reply-AVDwu3U")
     @Deprecated("For binary compatibility", level = DeprecationLevel.HIDDEN)
-    public infix fun MessageSelectionTimeoutChecker.reply000(message: Message): Void? {
+    public infix fun MessageSelectionTimeoutChecker.reply000(message: Message): Nothing? {
```
move into jvmMain

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/select.kt:
```diff
@@ -457,7 +460,7 @@ public abstract class MessageSelectBuilderUnit<M : MessageEvent, R> @PublishedAp
     @JvmName("quoteReply-RNyhSv4")
     @Deprecated("For binary compatibility", level = DeprecationLevel.HIDDEN)
-    public infix fun MessageSelectionTimeoutChecker.quoteReply000(block: suspend () -> Any?): Void? {
+    public infix fun MessageSelectionTimeoutChecker.quoteReply000(block: suspend () -> Any?): Nothing? {
```
move into jvmMain

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/select.kt:
```diff
@@ -470,7 +473,7 @@ public abstract class MessageSelectBuilderUnit<M : MessageEvent, R> @PublishedAp
     @JvmName("quoteReply-sCZ5gAI")
     @Deprecated("For binary compatibility", level = DeprecationLevel.HIDDEN)
-    public infix fun MessageSelectionTimeoutChecker.quoteReply000(message: String): Void? {
+    public infix fun MessageSelectionTimeoutChecker.quoteReply000(message: String): Nothing? {
```
move into jvmMain

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/event/select.kt:
```diff
@@ -483,7 +486,7 @@ public abstract class MessageSelectBuilderUnit<M : MessageEvent, R> @PublishedAp
     @JvmName("quoteReply-AVDwu3U")
     @Deprecated("For binary compatibility", level = DeprecationLevel.HIDDEN)
-    public infix fun MessageSelectionTimeoutChecker.quoteReply000(message: Message): Void? {
+    public infix fun MessageSelectionTimeoutChecker.quoteReply000(message: Message): Nothing? {
```
move into jvmMain

Him188 replied:
这个 class 可以考虑变 expect

AdoptOSS commented on mirai-core-api/src/commonMain/kotlin/message/data/Message.kt:
```diff
@@ -360,7 +360,7 @@ public inline fun Message.repeat(count: Int): MessageChain {
     }
     return buildMessageChain(count) {
-        repeat(count) {
+        repeat(count) l@{
```
why?

Him188 replied:
估计以前改的忘了删, 在这个 PR 里没必要修改

Him188 committed:
Configure apiBuild dependsOn jar tasks only

Him188 committed:
Upgrade to Ktor 2.0.2, use CIO engine for linux targets; Remove ktor-client-okhttp from mirai-core-api jvmBaseMain
Use Input.readAllText instead of Input.readText, because readText only reads one buffer.
#2084

Him188 committed:
core: Close and join all jobs after each test

Him188 commented:
使用 Kotlin 生成的 API 的 C++ 登录:

```c++
int main() {
    const auto *symbols = libmiraicore_symbols();
    const auto mirai = symbols->kotlin.root.net.mamoe.mirai;
    const auto BotConfiguration = mirai.utils.BotConfiguration;
    mirai.internal.initMirai();


    auto conf = BotConfiguration.BotConfiguration();

    BotConfiguration.set_workingDir(conf, ".");
    BotConfiguration.set_protocol(conf, BotConfiguration.MiraiProtocol.ANDROID_PAD.get());
    BotConfiguration.fileBasedDeviceInfo(conf, "device.local.json");

    auto bot = mirai.BotFactory.INSTANCE.newBot__(
            mirai.BotFactory.INSTANCE._instance(),
            123456L,
            my_passwrod,
            conf
    );

    mirai.Bot.loginB(bot);
    mirai.Bot.joinB(bot);
    return 0;
}
```

Him188 committed:
Produce shared libraries and static libraries

Him188 committed:
Remove MiraiLogger.subLogger in mirai-core-api to overcome declaration clash on native

Him188 committed:
Add RunConfigurations

Him188 committed:
Add dummy jvmBase target for IDE

Him188 committed:
Add docs for building core

Him188 committed:
Add test for `test Audio extraData`

Him188 committed:
Move AudioTest to message/data

Him188 committed:
Add docs for implementing new message

Him188 commented:
Actions Windows 平台 7GB 内存不够链接

Him188 committed:
Use kotlinx-coroutines-test

Him188 committed:
Fix compiler warnings

AdoptOSS commented:
尝试下通过调整GC策略来降低 RAM 占用，看看可不可行？
比如使用 OpenJ9 这种专注于 RAM 优化的 JVM

AdoptOSS commented:
尝试修改了一些actions的设置：
- OpenJ9可以减少内存使用，但似乎会导致jvm下2个test不通过
- Java占用内存过大可以修改jvmArgs为`-Xmx3072m`来减少异常，本来的`-Xmx7000m`会导致极大的内存占用（jvm不会过于积极地回收），由于link步骤需要调用外部程序，而外部程序的内存分配失败无法被Java的GC所感知（java自己分配的话，如果内存不足，会去跑一下回收程序再分配），过大的数值极易导致linker内存分配失败
- 在`-Xmx***`较低的情况下，甚至可以启用daemon（似乎build速度有所提升）
- Windows下native的编译配置，缺少安装curl的必要步骤（还没调试成功【我好菜】）

Him188 commented:
Xmx 4000，kotlin 编译器就会说  OOM，所以之前改成 6000 了，然后测试了一下 7000

AdoptOSS commented:
我这算是勉强在win下跑出了 lld: error: unable to find library -lcurl 吧
https://github.com/AdoptOSS/mirai/runs/6899864946

Sent from Outlook for Android<https://aka.ms/ghei36>

________________________________
发件人: Him188 ***@***.***>
发送时间: 2022年6月15日星期三 下午9:01
收件人: mamoe/mirai
抄送: AdoptOSS; Mention
主题: Re: [mamoe/mirai] Native target (PR #2048)


Xmx 4000，kotlin 编译器就会说 OOM，所以之前改成 6000 了，然后测试了一下 7000

―
Reply to this email directly, view it on GitHub<https://github.com/mamoe/mirai/pull/2048#issuecomment-1156443264>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AROIETEH2IE56BU37AVR2X3VPHH3ZANCNFSM5WOS6VAA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Him188 commented:
@AdoptOSS 尝试给 link 那个 task 单独 Xmx3000 可以跑出 error，目前在修库

AdoptOSS commented on mirai-core/src/nativeMain/cinterop/OpenSSL.def:
```diff
@@ -35,3 +37,5 @@ compilerOpts =               -I/opt/openssl/include \
                              -I/usr/local/include/ \
                              -IC:/openssl/include/ \
```
感觉要用 `-IC` 的吧

AdoptOSS commented on mirai-core/src/nativeMain/cinterop/OpenSSL.def:
```diff
@@ -35,3 +37,5 @@ compilerOpts =               -I/opt/openssl/include \
                              -I/usr/local/include/ \
                              -IC:/openssl/include/ \
                              -LC:/vcpkg/installed/x64-windows/include \
+                             -LC:/vcpkg/packages/curl_x64-windows/include \
```
感觉要用 `-IC` 的吧

Him188 replied:
确实

AdoptOSS commented:
把link也调试出来了
找不到库其实不是搜索目录的问题，而是Windows下的查找规则，不会自动给你加`lib`前缀
```yaml
      - if: ${{ env.isWindows == 'true' }}
        name: Install OpenSSL & cURL
        run: |
          echo "set(VCPKG_BUILD_TYPE release)" | Out-File -FilePath "$env:VCPKG_INSTALLATION_ROOT\triplets\x64-windows.cmake" -Encoding utf8 -Append
          vcpkg install openssl:x64-windows curl[core,openssl]:x64-windows
          New-Item -Path $env:VCPKG_INSTALLATION_ROOT\installed\x64-windows\lib\crypto.lib -ItemType SymbolicLink -Value $env:VCPKG_INSTALLATION_ROOT\installed\x64-windows\lib\libcrypto.lib
          New-Item -Path $env:VCPKG_INSTALLATION_ROOT\installed\x64-windows\lib\ssl.lib -ItemType SymbolicLink -Value $env:VCPKG_INSTALLATION_ROOT\installed\x64-windows\lib\libssl.lib
          New-Item -Path $env:VCPKG_INSTALLATION_ROOT\installed\x64-windows\lib\curl.lib -ItemType SymbolicLink -Value $env:VCPKG_INSTALLATION_ROOT\installed\x64-windows\lib\libcurl.lib
          echo "$env:VCPKG_INSTALLATION_ROOT\installed\x64-windows\bin" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
```

Him188 commented:
没 Windows 电脑了，最近一个月没法测了

Him188 committed:
Fix services

AdoptOSS committed:
native: 优化 build script 及 actions，修复 ECDH 相关资源释放问题 (#2110)

* build: fix build on Windows

* test: fix ContentEqualsTest on native

* build: allow disabling targets with property

* fix: free ECDH-related resources properly on native
avoid memory leaking (usually on *nix) or crash (usually on Windows)

Signed-off-by: AdoptOSS <adoptoss@outlook.com>

* build(workflow): remove invalid options

* fix(styling): `mirai.target` property

* build(workflow): try to limit memory usage during mirai-console:tools:gradle-plugin:integTest

* enhance(test): use buildList

* build(workflow): retry

Co-authored-by: ArcticLampyrid <arcticlampyrid@outlook.com>

Him188 committed:
Minor docs update for NetworkHandler

Him188 committed:
Configure shadow relocation and add checks for multiplatform publishing

AdoptOSS committed:
fix: platform socket on Windows (#2121)

* fix: platform socket on Windows

* fix(workflow): use win platform ssl provider instead of openssl for cURL
openssl can't use system's native CA store by default

Him188 committed:
actions publishing

Him188 committed:
Enable all targets when IDEA active

Him188 committed:
Disable JVM tests for Windows

Him188 committed:
Configure publishing for HMPP

Him188 committed:
Optimize gradle properties

Him188 committed:
Add multiplatform DynamicTest
